number_theory/modular.h

```diff
@@ -0,0 +1,82 @@
+#ifndef NUMBER_THEORY_MODULAR_H_
+#define NUMBER_THEORY_MODULAR_H_
+
+#include <stddef.h>
+
+#include <bit>
+#include <limits>
+
+#include "number_theory/utility.h"
+
+// Modular arithmetic
+
+namespace tql {
+namespace number_theory {
+
+template <class T, T mod>
+class Modular {
+  static_assert(std::numeric_limits<T>::is_integer && mod > 0,
+                "Modular requires modulus to be a positive integer.");
+
+ public:
+  using type = T;
+  static constexpr T modulus = mod;
+
+  Modular(type value = 0) { set(std::move(value)); }
+  Modular(const Modular &other) = default;
+  Modular(Modular &&other) = default;
+
+  Modular &operator=(const Modular &other) = default;
+  Modular &operator=(Modular &&other) = default;
+
+  const type &get() const { return value_; }
+
+  void set(type value) {
+    value_ = value % modulus;
+    if (value_ < 0)
+      value_ += modulus;
+  }
+
+  Modular add(const Modular &rhs) const {
+    check_addition_overflow();
+    return Modular((value_ + rhs.value_) % modulus);
```

Comment:
I think we should avoid `%` here such as:
```cpp
value_ += rhs.value_
if (value_ >= modulus)
  value_ -= modulus;
```
However, this might cause some race conditions since the write is not atomic and we might need some temporary variables.